go/vt/tabletserver/tabletserver.go

```diff
@@ -1367,27 +1375,75 @@ func splitQueryToTabletError(err error) error {
 	return NewTabletError(vtrpcpb.ErrorCode_BAD_INPUT, "splitquery: %v", err)
 }
 
-// StreamHealthRegister is part of queryservice.QueryService interface
-func (tsv *TabletServer) StreamHealthRegister(c chan<- *querypb.StreamHealthResponse) (int, error) {
+// StreamHealth is part of queryservice.QueryService interface
+func (tsv *TabletServer) StreamHealth(ctx context.Context, sendReply func(*querypb.StreamHealthResponse) error) error {
+	tsv.streamHealthMutex.Lock()
+	shr := tsv.lastStreamHealthResponse
+	tsv.streamHealthMutex.Unlock()
+	if shr != nil {
+		if err := sendReply(shr); err != nil {
+			if err == io.EOF {
+				return nil
+			}
+			return err
+		}
+	}
+	id, ch := tsv.streamHealthRegister()
+	defer tsv.streamHealthUnregister(id)
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case shr = <-ch:
+		}
+		if err := sendReply(shr); err != nil {
+			if err == io.EOF {
+				return nil
+			}
+			return err
+		}
+	}
+}
+
+func (tsv *TabletServer) streamHealthRegister() (id int, ch chan *querypb.StreamHealthResponse) {
 	tsv.streamHealthMutex.Lock()
 	defer tsv.streamHealthMutex.Unlock()
 
-	id := tsv.streamHealthIndex
+	id = tsv.streamHealthIndex
 	tsv.streamHealthIndex++
-	tsv.streamHealthMap[id] = c
-	if tsv.lastStreamHealthResponse != nil {
-		c <- tsv.lastStreamHealthResponse
-	}
-	return id, nil
+	ch = make(chan *querypb.StreamHealthResponse, 10)
+	tsv.streamHealthMap[id] = ch
+	return id, ch
```

Comment:
Yeah, I made that up but I think it makes sense.

As discussed offline, please do not use a named return variable here. The function is so short it will be clear from the `return` statement what the return values are :)